apps/distgo/cmd/dist/scheduler.go

```diff
@@ -0,0 +1,109 @@
+// Copyright 2016 Palantir Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package dist
+
+import (
+	"sort"
+
+	"github.com/pkg/errors"
+
+	"github.com/palantir/godel/apps/distgo/params"
+)
+
+func Schedule(specsWithDeps []params.ProductBuildSpecWithDeps) ([]params.ProductBuildSpecWithDeps, error) {
+	var schedule []params.ProductBuildSpecWithDeps
+	graph := make(map[string]map[string]bool)
+	specMap := make(map[string][]params.ProductBuildSpecWithDeps)
+	// create a graph of dependencies
+	for _, curSpec := range specsWithDeps {
+		product := curSpec.Spec.ProductName
+		if graph[product] == nil {
+			graph[product] = make(map[string]bool)
```

Comment:
we actually need this step because this creates a vertex in the graph even if there are no edges in the graph and this would result in the ordering to be of same length as the number of vertices in the input graph.